driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -7,177 +7,144 @@ use crate::{
-    // When this struct is dropped this sender will be dropped which makes the updater thread stop.
-    _exit_tx: oneshot::Sender<()>,
+    contract: Arc<dyn StableXContract + Send + Sync>,
+    web3: Web3,
+    /// We need a mutex because otherwise the struct wouldn't be Sync which is needed because we use
+    /// the orderbook in multiple threads. The mutex is locked in `get_auction_data` while the
+    /// orderbook is updated with new events.
+    behind_mutex: Mutex<BehindMutex>,
 }
 
-impl UpdatingOrderbook {
-    pub fn new(contract: Arc<dyn StableXContract + Send + Sync>, web3: Web3) -> Self {
-        let orderbook = Arc::new(Mutex::new(Orderbook::default()));
-        let orderbook_clone = orderbook.clone();
-        let orderbook_ready = Arc::new(AtomicBool::new(false));
-        let orderbook_ready_clone = orderbook_ready.clone();
-        let (exit_tx, exit_rx) = oneshot::channel();
+struct BehindMutex {
+    orderbook: Orderbook,
+    last_handled_block: u64,
+    block_timestamp_reader: CachedBlockTimestampReader<Web3>,
+}
 
-        std::thread::spawn(move || {
-            let result = futures::executor::block_on(update_with_events_forever(
-                web3.clone(),
-                contract,
-                orderbook_clone,
-                orderbook_ready_clone,
-                CachedBlockTimestampReader::new(web3),
-                exit_rx,
-            ));
-            if let Err(err) = result {
-                log::error!("event based orderbook failed: {:?}", err);
-                // TODO: implement a retry mechanism
-                // For now we error the program so force a restart of the whole driver because
-                // without a retry we would be stuck with an outdated orderbook forever.
-                // Sleep for one second, so that we have time to flush the logs.
-                thread::sleep(Duration::from_secs(1));
-                process::exit(1);
-            }
-        });
+impl UpdatingOrderbook {
+    /// Blocks on attempting to build the orderbook. After the initial build the orderbook is
+    /// updated every time `get_auction_data` is called.
+    pub fn new(contract: Arc<dyn StableXContract + Send + Sync>, web3: Web3) -> Result<Self> {
+        let result = Self {
+            contract,
+            web3: web3.clone(),
+            behind_mutex: Mutex::new(BehindMutex {
+                orderbook: Orderbook::default(),
+                last_handled_block: 0,
+                block_timestamp_reader: CachedBlockTimestampReader::new(web3),
+            }),
+        };
 
-        Self {
-            orderbook,
-            orderbook_ready,
-            _exit_tx: exit_tx,
+        // Perform the initial update.
+        {
+            let mut behind_mutex = result
+                .behind_mutex
+                .lock()
+                .map_err(|err| anyhow!("poison error: {}", err))?;
+            futures::executor::block_on(result.update_with_events(&mut behind_mutex))?;
```

Comment:
This is part of the initialization for the struct so I find it sensible.
I do like the idea of having the constructor be an async function instead of forcing the blocking. Then we can decide elsewhere how to handle the Future. Let's do that in a separate PR.

Reply:
I'm worried that in the current approach we will be stuck in the beginning of our main function for a minute or two while we are fetching the initial events. I'd really prefer if this happened already in the scheduler thread after all setup is done (otherwise we might panic after two minutes because some env variable is set incorrectly), rather than in main.

Reply:
Looking at `main.rs` I see that we are already creating the price finder before the orderbook so it isn't blocked and the command line options have already been parsed. The only thing I see that can still fail is creating the AuctionTimingConfiguration which we can move in front of the orderbook.

Reply:
I think generally we shouldn't have to think too much about the order in which we instantiate our components (other then explicit dependency satisfaction) and have to worry that some new functions might run for 2 minutes. 
I'd be fine with returning a promise and wait for it at the right time, but for now it seems the simplest to me to simply `get_auction_data()`in the scheduler to warm things up (this would also make it that you only access the inner Context in one place).